File: design/lsp/parser/ast/ast.go
package ast

// Node is the base interface for all AST nodes.
type Node interface {
	NodeLine() int
	NodeColumn() int
}

// Definition is a top-level definition (workflow or activity).
type Definition interface {
	Node
	defNode()
}

// Statement is a statement inside a body.
type Statement interface {
	Node
	stmtNode()
}

// Pos holds source position information.
type Pos struct {
	Line   int
	Column int
}

func (p Pos) NodeLine() int   { return p.Line }
func (p Pos) NodeColumn() int { return p.Column }

// File represents a parsed .twf file.
type File struct {
	Definitions []Definition
}

// ---------------------------------------------------------------------------
// Top-level definitions
// ---------------------------------------------------------------------------

type WorkflowDef struct {
	Pos
	Name       string
	Params     string // opaque content inside parens
	ReturnType string // opaque, optional
	Options    string // opaque, optional
	Signals    []*SignalDecl
	Queries    []*QueryDecl
	Updates    []*UpdateDecl
	Body       []Statement
}

func (*WorkflowDef) defNode() {}

type ActivityDef struct {
	Pos
	Name       string
	Params     string
	ReturnType string
	Options    string
	Body       []Statement
}

func (*ActivityDef) defNode() {}

// ---------------------------------------------------------------------------
// Workflow-level declarations (embedded in WorkflowDef)
// ---------------------------------------------------------------------------

type SignalDecl struct {
	Pos
	Name   string
	Params string
	Body   []Statement // handler body
}

func (*SignalDecl) stmtNode() {}

type QueryDecl struct {
	Pos
	Name       string
	Params     string
	ReturnType string
	Body       []Statement // handler body (restricted: no temporal primitives)
}

func (*QueryDecl) stmtNode() {}

type UpdateDecl struct {
	Pos
	Name       string
	Params     string
	ReturnType string
	Body       []Statement // handler body
}

func (*UpdateDecl) stmtNode() {}

// ---------------------------------------------------------------------------
// Statements
// ---------------------------------------------------------------------------

type ActivityCall struct {
	Pos
	Name     string
	Args     string
	Result   string // optional
	Options  string // optional
	Resolved *ActivityDef
}

func (*ActivityCall) stmtNode() {}

// WorkflowCallMode describes how a workflow call is executed.
type WorkflowCallMode int

const (
	CallChild  WorkflowCallMode = iota // bare workflow call (child)
	CallSpawn                          // spawn workflow
	CallDetach                         // detach workflow (fire-and-forget)
)

type WorkflowCall struct {
	Pos
	Mode      WorkflowCallMode
	Namespace string // optional, from nexus "ns"
	Name      string
	Args      string
	Result    string // optional
	Options   string // optional
	Resolved  *WorkflowDef
}

func (*WorkflowCall) stmtNode() {}

// AwaitStmt represents a single await statement.
type AwaitStmt struct {
	Pos
	// Timer await
	Timer string // duration, e.g. "5m"

	// Signal await
	Signal       string // signal name
	SignalParams string // optional parameter binding, e.g. "(approver, timestamp)"
	SignalResolved *SignalDecl

	// Update await
	Update       string // update name
	UpdateParams string // optional parameter binding
	UpdateResolved *UpdateDecl

	// Activity await
	Activity string // activity name
	ActivityArgs string
	ActivityResult string // optional result binding
	ActivityResolved *ActivityDef

	// Workflow await
	Workflow string // workflow name
	WorkflowMode WorkflowCallMode // spawn/detach
	WorkflowNamespace string // optional nexus namespace
	WorkflowArgs string
	WorkflowResult string // optional result binding
	WorkflowResolved *WorkflowDef
}

// AwaitKind returns the kind of await statement.
func (a *AwaitStmt) AwaitKind() string {
	switch {
	case a.Timer != "":
		return "timer"
	case a.Signal != "":
		return "signal"
	case a.Update != "":
		return "update"
	case a.Activity != "":
		return "activity"
	case a.Workflow != "":
		return "workflow"
	default:
		return "unknown"
	}
}

func (*AwaitStmt) stmtNode() {}

// AwaitAllBlock represents an "await all:" block that waits for all operations to complete.
type AwaitAllBlock struct {
	Pos
	Body []Statement
}

func (*AwaitAllBlock) stmtNode() {}

// AwaitOneCase represents a single case in an "await one:" block.
// Can be signal, update, timer, activity, workflow, or nested await all.
type AwaitOneCase struct {
	Pos

	// Signal case
	Signal string // signal name
	SignalParams string // optional parameter binding, e.g. "(approver, timestamp)"
	SignalResolved *SignalDecl

	// Update case
	Update string // update name
	UpdateParams string // optional parameter binding
	UpdateResolved *UpdateDecl

	// Timer case
	Timer string // duration

	// Activity case
	Activity string // activity name
	ActivityArgs string
	ActivityResult string // optional result binding
	ActivityResolved *ActivityDef

	// Workflow case
	Workflow string // workflow name
	WorkflowMode WorkflowCallMode // spawn/detach/child
	WorkflowNamespace string // optional nexus namespace
	WorkflowArgs string
	WorkflowResult string // optional result binding
	WorkflowResolved *WorkflowDef

	// Await all case (nested)
	AwaitAll *AwaitAllBlock

	Body []Statement // optional body (can be empty)
}

// CaseKind returns the kind of this await one case.
func (c *AwaitOneCase) CaseKind() string {
	switch {
	case c.Signal != "":
		return "signal"
	case c.Update != "":
		return "update"
	case c.Timer != "":
		return "timer"
	case c.Activity != "":
		return "activity"
	case c.Workflow != "":
		return "workflow"
	case c.AwaitAll != nil:
		return "await_all"
	default:
		return "unknown"
	}
}

func (*AwaitOneCase) stmtNode() {}

// AwaitOneBlock represents an "await one:" block that waits for the first case to complete.
type AwaitOneBlock struct {
	Pos
	Cases []*AwaitOneCase
}

func (*AwaitOneBlock) stmtNode() {}

// SwitchCase represents a single case in a switch block.
type SwitchCase struct {
	Pos
	Value string // opaque expression after "case"
	Body  []Statement
}

type SwitchBlock struct {
	Pos
	Expr    string // opaque, paren-delimited
	Cases   []*SwitchCase
	Default []Statement // optional else block
}

func (*SwitchBlock) stmtNode() {}

type IfStmt struct {
	Pos
	Condition string // opaque, paren-delimited
	Body      []Statement
	ElseBody  []Statement // optional
}

func (*IfStmt) stmtNode() {}

// ForVariant describes the kind of for loop.
type ForVariant int

const (
	ForInfinite    ForVariant = iota // for:
	ForConditional                   // for (condition):
	ForIteration                     // for (var in collection):
)

type ForStmt struct {
	Pos
	Variant   ForVariant
	Condition string // for conditional loops
	Variable  string // for iteration loops
	Iterable  string // for iteration loops
	Body      []Statement
}

func (*ForStmt) stmtNode() {}

type ReturnStmt struct {
	Pos
	Value string // opaque, optional
}

func (*ReturnStmt) stmtNode() {}

type CloseStmt struct {
	Pos
	Reason string // "completed", "failed", or "" (default is completed)
	Value  string // opaque, optional
}

func (*CloseStmt) stmtNode() {}

type ContinueAsNewStmt struct {
	Pos
	Args string
}

func (*ContinueAsNewStmt) stmtNode() {}

type BreakStmt struct {
	Pos
}

func (*BreakStmt) stmtNode() {}

type ContinueStmt struct {
	Pos
}

func (*ContinueStmt) stmtNode() {}

type RawStmt struct {
	Pos
	Text string
}

func (*RawStmt) stmtNode() {}

type Comment struct {
	Pos
	Text string
}

func (*Comment) stmtNode() {}

File: design/visualizer/src/types/ast.ts
// TypeScript types mirroring the Go AST JSON output

export interface Position {
  line: number
  column: number
}

// Per-file parse error
export interface FileError {
  file: string
  error: string
  stderr?: string
}

// Top-level file
export interface TWFFile {
  definitions: Definition[]
  // Added for focused-file visualization
  focusedFile?: string
  // Per-file parse errors and warnings
  errors?: FileError[]
}

// Definition types
export type Definition = WorkflowDef | ActivityDef

export interface WorkflowDef extends Position {
  type: 'workflowDef'
  name: string
  params: string
  returnType?: string
  options?: string
  signals: SignalDecl[]
  queries: QueryDecl[]
  updates: UpdateDecl[]
  body: Statement[]
  // Source file path (added by extension)
  sourceFile?: string
}

export interface ActivityDef extends Position {
  type: 'activityDef'
  name: string
  params: string
  returnType?: string
  options?: string
  body: Statement[]
  // Source file path (added by extension)
  sourceFile?: string
}

// Declaration types (with handler bodies)
export interface SignalDecl extends Position {
  type: 'signalDecl'
  name: string
  params: string
  body?: Statement[]
}

export interface QueryDecl extends Position {
  type: 'queryDecl'
  name: string
  params: string
  returnType?: string
  body?: Statement[]
}

export interface UpdateDecl extends Position {
  type: 'updateDecl'
  name: string
  params: string
  returnType?: string
  body?: Statement[]
}

// Statement types
export type Statement =
  | ActivityCall
  | WorkflowCall
  | AwaitStmt
  | AwaitAllBlock
  | AwaitOneBlock
  | SwitchBlock
  | IfStmt
  | ForStmt
  | ReturnStmt
  | CloseStmt
  | ContinueAsNewStmt
  | BreakStmt
  | ContinueStmt
  | RawStmt
  | Comment

export interface ActivityCall extends Position {
  type: 'activityCall'
  name: string
  args: string
  result?: string
  options?: string
}

export type WorkflowCallMode = 'child' | 'spawn' | 'detach'

export interface WorkflowCall extends Position {
  type: 'workflowCall'
  mode: WorkflowCallMode
  namespace?: string
  name: string
  args: string
  result?: string
  options?: string
}

// Single await statement: await timer/signal/update/activity/workflow
export type AwaitStmtKind = 'timer' | 'signal' | 'update' | 'activity' | 'workflow'

export interface AwaitStmt extends Position {
  type: 'await'
  kind: AwaitStmtKind
  timer?: string
  signal?: string
  signalParams?: string
  update?: string
  updateParams?: string
  activity?: string
  activityArgs?: string
  activityResult?: string
  workflow?: string
  workflowMode?: string
  workflowNamespace?: string
  workflowArgs?: string
  workflowResult?: string
}

// await all: waits for all operations to complete
export interface AwaitAllBlock extends Position {
  type: 'awaitAll'
  body: Statement[]
}

// await one case: signal, update, timer, activity, workflow, or nested await all
export type AwaitOneCaseKind = 'signal' | 'update' | 'timer' | 'activity' | 'workflow' | 'await_all'

export interface AwaitOneCase {
  kind: AwaitOneCaseKind
  // Signal case
  signal?: string
  signalParams?: string
  // Update case
  update?: string
  updateParams?: string
  // Timer case
  timer?: string
  // Activity case
  activity?: string
  activityArgs?: string
  activityResult?: string
  // Workflow case
  workflow?: string
  workflowMode?: string
  workflowNamespace?: string
  workflowArgs?: string
  workflowResult?: string
  // Await all case (nested)
  awaitAll?: AwaitAllBlock
  // Body executed when this case wins (optional - can be empty)
  body: Statement[]
}

// await one: waits for first case to complete
export interface AwaitOneBlock extends Position {
  type: 'awaitOne'
  cases: AwaitOneCase[]
}

export interface SwitchCase {
  value: string
  body: Statement[]
}

export interface SwitchBlock extends Position {
  type: 'switch'
  expr: string
  cases: SwitchCase[]
  default?: Statement[]
}

export interface IfStmt extends Position {
  type: 'if'
  condition: string
  body: Statement[]
  elseBody?: Statement[]
}

export type ForVariant = 'infinite' | 'conditional' | 'iteration'

export interface ForStmt extends Position {
  type: 'for'
  variant: ForVariant
  condition?: string
  variable?: string
  iterable?: string
  body: Statement[]
}

export interface ReturnStmt extends Position {
  type: 'return'
  value?: string
}

export interface CloseStmt extends Position {
  type: 'close'
  reason?: string // 'completed', 'failed', or empty (default completed)
  value?: string
}

export interface ContinueAsNewStmt extends Position {
  type: 'continueAsNew'
  args: string
}

export interface BreakStmt extends Position {
  type: 'break'
}

export interface ContinueStmt extends Position {
  type: 'continue'
}

export interface RawStmt extends Position {
  type: 'raw'
  text: string
}

export interface Comment extends Position {
  type: 'comment'
  text: string
}

// Type guards
export function isWorkflowDef(def: Definition): def is WorkflowDef {
  return def.type === 'workflowDef'
}

export function isActivityDef(def: Definition): def is ActivityDef {
  return def.type === 'activityDef'
}
